### PR TITLE
[Tab Switch]: Prefer switching to an existing tab over navigating to the same url in a new one

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/32896

Without flag:
image

With flag:
image

The plan is to enable this in Nightly for a while (via Griffin) and see what people think

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. Enable brave://flags/#brave-omnibox-tab-switch-by-default
2. Enter a currently open tab (for some reason, not all open tabs show the switcher, I'm not sure why, but it comes from upstream)
3. The URL bar should switch you to the existing tab by default, rather than forcing you to tab

### CODE REVIEW

File: components/omnibox/common/brave_omnibox_features.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_COMPONENTS_OMNIBOX_COMMON_BRAVE_OMNIBOX_FEATURES_H_
+#define BRAVE_COMPONENTS_OMNIBOX_COMMON_BRAVE_OMNIBOX_FEATURES_H_
+
+#include "base/feature_list.h"
+
+namespace omnibox {
+
+BASE_DECLARE_FEATURE(kOmniboxTabSwitchByDefault);
```
Comment (goodov) [inline components/omnibox/common/brave_omnibox_features.h:13]: pls add features into `components/omnibox/common/omnibox_features.h` via `chromium_src` override. This is a preferred way when you need to reference them from upstream files or in general use them closely with the upstream logic.

File: vector_icons/leo_overrides.json [hunk 2]
```diff
@@ -23,5 +23,6 @@
   "//components/vector_icons/forward_arrow.icon": "leo_browser_forward",
   "//components/vector_icons/https_valid.icon": "leo_lock_plain",
   "//components/vector_icons/not_secure_warning.icon": "leo_warning_triangle_filled",
-  "//components/vector_icons/reload.icon": "leo_browser_refresh"
+  "//components/vector_icons/reload.icon": "leo_browser_refresh",
+  "//components/omnibox/browser/vector_icons/switch.icon": "leo_window_tab"
```
Comment (goodov) [inline vector_icons/leo_overrides.json:27]: sort

File: browser/about_flags.cc [hunk 3]
```diff
@@ -409,6 +410,14 @@
 #define BRAVE_AI_CHAT
 #endif
 
+#define BRAVE_OMNIBOX_FEATURES                                               \
+  EXPAND_FEATURE_ENTRIES(                                                    \
+      {"brave-omnibox-tab-switch-by-default", "Brave Tab Switch by Default", \
+       "Prefer switching to already open tabs, rather than navigating in a " \
+       "new tab",                                                            \
+       kOsWin | kOsLinux | kOsMac,                                           \
+       FEATURE_VALUE_TYPE(omnibox::kOmniboxTabSwitchByDefault)})
```
Comment (goodov) [inline browser/about_flags.cc:419]: pls add a trailing comma to format this a little better.

File: chromium_src/components/omnibox/browser/DEPS [hunk 4]
```diff
@@ -1,7 +1,7 @@
 include_rules = [
   "+brave/components/brave_search_conversion",
   "+brave/components/commander",
-  "+brave/components/omnibox/browser",
+  "+brave/components/omnibox",
```
Comment (goodov) [inline chromium_src/components/omnibox/browser/DEPS:4]: this should be reverted when you move features into chromium_src override.

File: components/omnibox/common/brave_omnibox_features.h (hunk 1 above)
Comment (fallaciousreasoning) [reply components/omnibox/common/brave_omnibox_features.h]: Cool, that makes sense!

**Review by sangwoo108 — APPROVED** [review]
LGTM!

**Comment by mkarolin** [comment]
I find the upstream UI counter-intuitive with the pill needing an additional selection when the row is already selected. But, I am not sure about this alternative either. `Switch to this tab` text gets partially bolded based on the search term length, which is confusing and we lose the actual URL. 

This might be too much work, but I would personally prefer if we kept the URL and instead have 2 pills. One pill could be `Switch to this tab` and the other `Open again in current tab`. `Switch to this tab` would be selected by default. It would also be nice to have an icon for the entire row indicating that this suggestion is an already opened tab instead of having the favicon there (or a combination of the two).

Maybe the design team can chime in. cc: @rebron

File: components/resources/omnibox_strings.grdp [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    Open here
+  </message>
+  <message name="IDS_ACC_OPEN_HERE_SUFFIX"
+    desc="Suffix for open here suggestions to explain keystroke used to open here.">
+    <ph name="TAB_SWITCH_SUFFIX">$1<ex>Brave http://brave.com bookmark</ex></ph>, currently open,
+    press Tab then Enter to switch to the open tab </message>
+  <message name="IDS_ACC_OPEN_HERE_BUTTON" desc="Announcement when open here button focused.">
+    Open here button, press Enter to open this url in the current tab, instead of switching to the
+    existing tab
+  </message>
```
Comment (mkarolin) [inline components/resources/omnibox_strings.grdp:14]: Are these switched? `Enter` should be to switch to the open tab and `Tab + Enter` to open here, no?
Also, maybe we should say `currently open in another tab`?

Comment (fallaciousreasoning) [reply components/resources/omnibox_strings.grdp:14]: I think these are correct:
1. `IDS_ACC_OPEN_HERE_SUFFIX` is a suffix for the AutocompleteMatch, which tells the user how to open the result in the current tab, if they want to (press tab, to select the pill, then press enter)
2. `IDS_ACC_OPEN_HERE_BUTTON` is the text if the button is focused. It explains what to do, and how to trigger it.

**Comment by fallaciousreasoning** [comment]
cc @aguscruiz for design input

> I find the upstream UI counter-intuitive with the pill needing an additional selection when the row is already selected

Agreed, it's definitely not ideal

> `Switch to this tab text` gets partially bolded based on the search term length, which is confusing and we lose the actual URL.

Whoops - this was a bug, I wasn't resetting the `contents_class`. Should be fixed now. Losing the actual URL is okay, I think. Bookmarks works similarly

image

> but I would personally prefer if we kept the URL and instead have 2 pills

I think that'd probably be better too - the massive pill with the text saying what isn't going to happen is pretty confusing. However, I think this is better than what we have currently, so would you be okay to land as is, and wait for @aguscruiz to have a look before doing something like that?

> nice to have an icon for the entire row indicating that this suggestion

Agreed, that'd be great! Ideally, I think it'd be the favicon, badged with a little launch arrow

**Review by mkarolin — APPROVED** [review]
LGTM

**Comment by fallaciousreasoning** [comment]
Going to land as is and see what UX thinks :smile:

**Pull request merged**